core/chains/evm/log/registrations.go

```diff
@@ -196,80 +269,117 @@ func filtersContainValues(topicValues []common.Hash, filters [][]Topic) bool {
+					}
+				}
+				r.logger.Tracef("No sub exists for addr %s and topic %s at this or greater MinIncomingConfirmations of %v. Resubscribe is required", addr.Hex(), topic.Hex(), sub.opts.MinIncomingConfirmations)
+				needsResubscribe = true
+			}
 		}
+	TopicSubscribedAtGreaterConf:
+		r.lookupSubs[addr][topic][sub] = topicValueFilters
 	}
 	return
 }
 
-func (r *handlers) removeSubscriber(sub *subscriber) (needsResubscribe bool) {
+func (r *handler) removeSubscriber(sub *subscriber, allHandlers map[uint32]*handler) (needsResubscribe bool) {
 	addr := sub.opts.Contract
 
-	// FIXME: What about the case where you remove/add a job with the same contract address?
-	// addr is not good enough to be a unique key
-	// MARK MARK MARK
-	if _, exists := r.handlersByAddr[addr]; !exists {
-		return
-	}
 	for topic := range sub.opts.LogsWithTopics {
-		topicMap, exists := r.handlersByAddr[addr][topic]
+		// OK to panic on missing addr/topic here, since that would be an invariant violation:
+		// Both addr and topic will always have been added on addSubscriber
+		// LogsWithTopics should never be mutated
+		// Only removeSubscriber should ever remove anything from this map
+		addrTopics, exists := r.lookupSubs[addr]
 		if !exists {
-			continue
+			r.logger.Panicf("AssumptionViolation: expected lookupSubs to contain addr %s for subscriber %p with job ID %v", addr.Hex(), sub, sub.listener.JobID())
 		}
+		topicMap, exists := addrTopics[topic]
+		if !exists {
+			r.logger.Panicf("AssumptionViolation: expected addrTopics to contain topic %v for subscriber %p with job ID %v", topic, sub, sub.listener.JobID())
+		}
+		if _, exists = topicMap[sub]; !exists {
+			r.logger.Panicf("AssumptionViolation: expected topicMap to contain subscriber %p with job ID %v", sub, sub.listener.JobID())
+		}
+		delete(topicMap, sub)
 
-		delete(topicMap, sub.listener)
-
+		// cleanup and resubscribe if necessary
 		if len(topicMap) == 0 {
-			needsResubscribe = true
-			delete(r.handlersByAddr[addr], topic)
+			r.logger.Tracef("No subs left for addr %s and topic %s at this MinIncomingConfirmations of %v", addr.Hex(), topic.Hex(), sub.opts.MinIncomingConfirmations)
+
+			if !needsResubscribe {
+				// NOTE: This is an optimization. Resub not necessary if there
+				// are still any other handlers listening on this addr/topic.
+				for confs, otherHandler := range allHandlers {
+					if confs == sub.opts.MinIncomingConfirmations {
+						// no need to check ourself, already did this above
+						continue
+					}
+					if _, exists := otherHandler.lookupSubs[addr][topic]; exists {
+						r.logger.Tracef("Sub still exists for addr %s and topic %s. Resubscribe will not be performed", addr.Hex(), topic.Hex())
+						goto TopicStillSubscribedByOtherHandler
+					}
+				}
+
+				r.logger.Tracef("No sub exists for addr %s and topic %s. Resubscribe will be performed", addr.Hex(), topic.Hex())
+				needsResubscribe = true
+			}
+		TopicStillSubscribedByOtherHandler:
```

Comment:
Is this equivalent, to avoid the `goto`? Maybe even cleaner as a separate helper.
```suggestion
			if !needsResubscribe {
			        func() {
				        // NOTE: This is an optimization. Resub not necessary if there
				        // are still any other handlers listening on this addr/topic.
				        for confs, otherHandler := range allHandlers {
					        if confs == sub.opts.MinIncomingConfirmations {
						        // no need to check ourself, already did this above
						        continue
					        }
					        if _, exists := otherHandler.lookupSubs[addr][topic]; exists {
						        r.logger.Tracef("Sub still exists for addr %s and topic %s. Resubscribe will not be performed", addr.Hex(), topic.Hex())
						        return
					        }
				        }
        
				        r.logger.Tracef("No sub exists for addr %s and topic %s. Resubscribe will be performed", addr.Hex(), topic.Hex())
				        needsResubscribe = true
				}()
			}
```